node/src/api/dto.rs

```diff
@@ -0,0 +1,186 @@
+    pub size: u64, // size in bytes of the encoded tx
+}
+
+/// Description of the current blockchain state.
+#[derive(Serialize)]
+pub struct StateDTO {
+    // Block header
+    pub tip: BlockHeaderDTO,
+    // The utreexo state
+    #[serde(with = "BigArray")]
+    pub utreexo: [Option<[u8; 32]>; 64],
+}
+
+/// Description of a connected peer.
+#[derive(Serialize)]
+pub struct PeerDTO {
+    pub id: [u8; 32],
+    pub since: u64,
+    /// ipv6 address format
+    pub addr: [u8; 16],
+    pub priority: u64,
+}
+
+#[derive(Serialize)]
+pub enum AnnotatedActionDTO {
+    Issue(IssueActionDTO),
+    Spend(SpendActionDTO),
+    Receive(ReceiveActionDTO),
+    Retire(RetireActionDTO),
+    Memo(MemoActionDTO),
+}
+
+#[derive(Serialize)]
+pub struct IssueActionDTO {
+    // Index of the txlog entry
+    pub entry: u32,
+    pub qty: u64,
+    pub flv: [u8; 32],
+}
+
+#[derive(Serialize)]
+pub struct SpendActionDTO {
+    // Index of the txlog entry
+    pub entry: u32,
+    pub qty: u64,
+    pub flv: [u8; 32],
+    // Identifier of the account sending funds
+    pub account: [u8; 32],
+}
+
+#[derive(Serialize)]
+pub struct ReceiveActionDTO {
+    // Index of the txlog entry
+    pub entry: u32,
+    pub qty: u64,
+    pub flv: [u8; 32],
+    // Identifier of the account receiving funds (if known)
+    pub account: Option<[u8; 32]>,
+}
+
+#[derive(Serialize)]
+pub struct RetireActionDTO {
+    // Index of the txlog entry
+    pub entry: u32,
+    pub qty: u64,
+    pub flv: [u8; 32],
+}
+
+#[derive(Serialize)]
+pub struct MemoActionDTO {
+    pub entry: u32,
+    pub data: Vec<u8>,
+}
+
+/// Description of the current blockchain state.
+#[derive(Serialize)]
+pub struct AnnotatedTxDTO {
+    /// Raw tx
+    pub tx: TxDTO,
+    pub actions: Vec<AnnotatedActionDTO>,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub enum BuildTxActionDTO {
+    IssueToAddress([u8; 32], u64, String),
+    IssueToReceiver(Receiver),
+    TransferToAddress([u8; 32], u64, String),
+    TransferToReceiver(Receiver),
+    Memo(Vec<u8>),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Cursor {
+    pub cursor: String,
+    pub count: Option<u32>,
+}
+
+impl Cursor {
+    const DEFAULT_ELEMENTS_PER_PAGE: u32 = 20;
+    pub fn count(&self) -> u32 {
+        self.count.unwrap_or(Self::DEFAULT_ELEMENTS_PER_PAGE)
+    }
+}
+
+#[derive(Deserialize)]
+pub struct HexId([u8; 32]);
+
+impl FromStr for HexId {
+    type Err = serde_json::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let array = serde_json::from_str(s)?;
+        Ok(Self(array))
+    }
+}
+
+impl From<blockchain::BlockHeader> for BlockHeaderDTO {
+    fn from(header: blockchain::BlockHeader) -> Self {
+        let blockchain::BlockHeader { version, height, prev, timestamp_ms, txroot, utxoroot, ext } = header;
+        Self {
+            version,
+            height,
+            prev: prev.0,
+            timestamp_ms,
+            txroot: txroot.0,
```

Comment:
maybe use the Hash type that we already have?

Reply:
It is hard question about newtypes. API users should not be aware of what newtypes we use inside the application. But using them will make code a little easier because they are used internally by the application.